File: bot-api/java/src/test/java/test_utils/MockedServer.java
package test_utils;

import com.google.gson.Gson;
import dev.robocode.tankroyale.schema.*;
import org.java_websocket.WebSocket;
import org.java_websocket.handshake.ClientHandshake;
import org.java_websocket.server.WebSocketServer;

import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.Set;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static dev.robocode.tankroyale.schema.Message.Type.*;

public final class MockedServer {

    public static final int PORT = 7913;

    public static String SESSION_ID = "123abc";
    public static String NAME = MockedServer.class.getSimpleName();
    public static String VERSION = "1.0.0";
    public static String VARIANT = "Tank Royale";
    public static Set<String> GAME_TYPES = Set.of("melee", "classic", "1v1");
    public static int MY_ID = 1;
    public static String GAME_TYPE = "classic";
    public static int ARENA_WIDTH = 800;
    public static int ARENA_HEIGHT = 600;
    public static int NUMBER_OF_ROUNDS = 10;
    public static double GUN_COOLING_RATE = 0.1;
    public static int MAX_INACTIVITY_TURNS = 450;
    public static int TURN_TIMEOUT = 30_000;
    public static int READY_TIMEOUT = 1_000_000;

    public static int BOT_ENEMY_COUNT = 7;
    public static double BOT_ENERGY = 99.7;
    public static double BOT_X = 44.5;
    public static double BOT_Y = 721.34;
    public static double BOT_DIRECTION = 120.1;
    public static double BOT_GUN_DIRECTION = 3.45;
    public static double BOT_RADAR_DIRECTION = 653.3;
    public static double BOT_RADAR_SWEEP = 13.5;
    public static double BOT_SPEED = 8.0;
    public static double BOT_TURN_RATE = 5.1;
    public static double BOT_GUN_TURN_RATE = 18.9;
    public static double BOT_RADAR_TURN_RATE = 34.1;
    public static double BOT_GUN_HEAT = 7.6;

    private double botEnergy = BOT_ENERGY;
    private double botGunHeat = BOT_GUN_HEAT;

    private final WebSocketServerImpl server = new WebSocketServerImpl();

    private final CountDownLatch openedLatch = new CountDownLatch(1);
    private final CountDownLatch botHandshakeLatch = new CountDownLatch(1);
    private final CountDownLatch gameStartedLatch = new CountDownLatch(1);
    private final CountDownLatch tickEventLatch = new CountDownLatch(1);
    private final CountDownLatch botIntentLatch = new CountDownLatch(1);

    private CountDownLatch botIntentContinueLatch = new CountDownLatch(1);

    private final Gson gson = new Gson();

    private BotHandshake botHandshake;
    private BotIntent botIntent;

    private int turnNumber = 1;


    public static URI getServerUrl() {
        try {
            return new URI("ws://localhost:" + PORT);
        } catch (URISyntaxException ex) {
            throw new RuntimeException(ex);
        }
    }

    public void start() {
        server.start();
    }

    public void stop() {
        try {
            server.stop();
            Thread.sleep(100);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void setBotEnergy(double botEnergy) {
        this.botEnergy = botEnergy;
    }

    public void setBotGunHeat(double botGunHeat) {
        this.botGunHeat = botGunHeat;
    }

    public boolean awaitConnection(int milliSeconds) {
        try {
            return openedLatch.await(milliSeconds, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            System.err.println("awaitConnection() was interrupted");
        }
        return false;
    }

    public boolean awaitBotHandshake(int milliSeconds) {
        try {
            return botHandshakeLatch.await(milliSeconds, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            System.err.println("awaitBotHandshake() was interrupted");
        }
        return false;
    }

    public boolean awaitGameStarted(int milliSeconds) {
        try {
            return gameStartedLatch.await(milliSeconds, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            System.err.println("awaitGameStarted() was interrupted");
        }
        return false;
    }

    public boolean awaitTickEvent(int milliSeconds) {
        try {
            return tickEventLatch.await(milliSeconds, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            System.err.println("awaitTickEvent() was interrupted");
        }
        return false;
    }

    public boolean awaitBotIntent(int milliSeconds) {
        try {
            botIntentContinueLatch.countDown();
            return botIntentLatch.await(milliSeconds, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            System.err.println("awaitBotIntent() was interrupted");
        }
        return false;
    }

    public BotHandshake getBotHandshake() {
        return botHandshake;
    }

    private class WebSocketServerImpl extends WebSocketServer {

        public WebSocketServerImpl() {
            super(new InetSocketAddress(PORT));
        }

        @Override
        public void onStart() {
        }

        @Override
        public void onOpen(WebSocket conn, ClientHandshake handshake) {
            openedLatch.countDown();
            sendServerHandshake(conn);
        }

        @Override
        public void onClose(WebSocket conn, int code, String reason, boolean remote) {
        }

        @Override
        public void onMessage(WebSocket conn, String text) {
            var message = gson.fromJson(text, Message.class);
            switch (message.getType()) {
                case BOT_HANDSHAKE:
//                    System.out.println("BOT_HANDSHAKE");

                    botHandshake = gson.fromJson(text, BotHandshake.class);
                    botHandshakeLatch.countDown();

                    sendGameStartedForBot(conn);
                    gameStartedLatch.countDown();
                    break;

                case BOT_READY:
//                    System.out.println("BOT_READY");

                    sendRoundStarted(conn);

                    sendTickEventForBot(conn, turnNumber++);
                    tickEventLatch.countDown();
                    break;

                case BOT_INTENT:
//                    System.out.println("BOT_INTENT");

                    try {
                        botIntentContinueLatch.await();
                    } catch (InterruptedException e) {
                        throw new RuntimeException(e);
                    }
                    botIntentContinueLatch = new CountDownLatch(1);

                    botIntent = gson.fromJson(text, BotIntent.class);
                    botIntentLatch.countDown();

                    sendTickEventForBot(conn, turnNumber++);
                    tickEventLatch.countDown();
                    break;
            }
        }

        @Override
        public void onMessage(WebSocket conn, ByteBuffer message) {
        }

        @Override
        public void onError(WebSocket conn, Exception ex) {
            throw new IllegalStateException("MockedServer error", ex);
        }

        private void sendServerHandshake(WebSocket conn) {
            var serverHandshake = new ServerHandshake();
            serverHandshake.setType(SERVER_HANDSHAKE);
            serverHandshake.setSessionId(SESSION_ID);
            serverHandshake.setName(NAME);
            serverHandshake.setVersion(VERSION);
            serverHandshake.setVariant(VARIANT);
            serverHandshake.setGameTypes(GAME_TYPES);
            send(conn, serverHandshake);
        }

        private void sendGameStartedForBot(WebSocket conn) {
            var gameStarted = new GameStartedEventForBot();
            gameStarted.setType(GAME_STARTED_EVENT_FOR_BOT);
            gameStarted.setMyId(MY_ID);
            var gameSetup = new GameSetup();
            gameSetup.setGameType(GAME_TYPE);
            gameSetup.setArenaWidth(ARENA_WIDTH);
            gameSetup.setArenaHeight(ARENA_HEIGHT);
            gameSetup.setNumberOfRounds(NUMBER_OF_ROUNDS);
            gameSetup.setGunCoolingRate(GUN_COOLING_RATE);
            gameSetup.setMaxInactivityTurns(MAX_INACTIVITY_TURNS);
            gameSetup.setTurnTimeout(TURN_TIMEOUT);
            gameSetup.setReadyTimeout(READY_TIMEOUT);
            gameStarted.setGameSetup(gameSetup);
            send(conn, gameStarted);
        }

        private void sendRoundStarted(WebSocket conn) {
            var roundStarted = new RoundStartedEvent();
            roundStarted.setType(ROUND_STARTED_EVENT);
            roundStarted.setRoundNumber(1);
            send(conn, roundStarted);
        }

        private void sendTickEventForBot(WebSocket conn, int turnNumber) {
            var tickEvent = new TickEventForBot();
            tickEvent.setType(TICK_EVENT_FOR_BOT);
            tickEvent.setRoundNumber(1);
            tickEvent.setTurnNumber(turnNumber);
            tickEvent.setEnemyCount(BOT_ENEMY_COUNT);

            var state = new BotState();
            state.setEnergy(botEnergy);
            state.setX(BOT_X);
            state.setY(BOT_Y);
            state.setDirection(BOT_DIRECTION);
            state.setGunDirection(BOT_GUN_DIRECTION);
            state.setRadarDirection(BOT_RADAR_DIRECTION);
            state.setRadarSweep(BOT_RADAR_SWEEP);
            state.setSpeed(BOT_SPEED);
            state.setTurnRate(BOT_TURN_RATE);
            state.setGunTurnRate(BOT_GUN_TURN_RATE);
            state.setRadarTurnRate(BOT_RADAR_TURN_RATE);
            state.setGunHeat(botGunHeat);
            tickEvent.setBotState(state);

            if (botIntent != null) {
                if (botIntent.getTurnRate() != null) {
                    state.setTurnRate(botIntent.getTurnRate());
                }
                if (botIntent.getGunTurnRate() != null) {
                    state.setGunTurnRate(botIntent.getGunTurnRate());
                }
                if (botIntent.getRadarTurnRate() != null) {
                    state.setRadarTurnRate(botIntent.getRadarTurnRate());
                }
            }

            var bulletState1 = createBulletState(1);
            var bulletState2 = createBulletState(2);
            tickEvent.setBulletStates(List.of(bulletState1, bulletState2));

            var event = new ScannedBotEvent();
            event.setType(SCANNED_BOT_EVENT);
            event.setDirection(45.0);
            event.setX(134.56);
            event.setY(256.7);
            event.setEnergy(56.9);
            event.setSpeed(9.6);
            event.setTurnNumber(1);
            event.setScannedBotId(2);
            event.setScannedByBotId(1);
            tickEvent.setEvents(List.of(event));

            send(conn, tickEvent);
        }

        private void send(WebSocket conn, Message message) {
            conn.send(gson.toJson(message));
        }

        private BulletState createBulletState(int id) {
            var bulletState = new BulletState();
            bulletState.setBulletId(id);
            bulletState.setX(0.0);
            bulletState.setY(0.0);
            bulletState.setOwnerId(0);
            bulletState.setDirection(0.0);
            bulletState.setPower(0.0);
            return bulletState;
        }
    }
}

File: bot-api/dotnet/Robocode.TankRoyale.BotApi.Tests/src/test_utils/MockedServer.cs
using System;
using System.Collections.Generic;
using System.Threading;
using Fleck;
using Newtonsoft.Json;
using Robocode.TankRoyale.BotApi.Util;
using Robocode.TankRoyale.Schema;

namespace Robocode.TankRoyale.BotApi.Tests.Test_utils;

public class MockedServer
{
    public const int Port = 7913;

    public static string SessionId = "123abc";
    public static string Name = nameof(MockedServer);
    public static string Version = "1.0.0";
    public static string Variant = "Tank Royale";
    public static ISet<string> GameTypes = new HashSet<string> { "melee", "classic", "1v1" };
    public static int MyId = 1;
    public static string GameType = "classic";
    public static int ArenaWidth = 800;
    public static int ArenaHeight = 600;
    public static int NumberOfRounds = 10;
    public static double GunCoolingRate = 0.1;
    public static int MaxInactivityTurns = 450;
    public static int TurnTimeout = 30_000;
    public static int ReadyTimeout = 1_000_000;

    public static int BotEnemyCcount = 7;
    public static double BotEnergy = 99.7;
    public static double BotX = 44.5;
    public static double BotY = 721.34;
    public static double BotDirection = 120.1;
    public static double BotGunDirection = 3.45;
    public static double BotRadarDirection = 653.3;
    public static double BotRadarSweep = 13.5;
    public static double BotSpeed = 8.0;
    public static double BotTurnRate = 5.1;
    public static double BotGunTurnRate = 18.9;
    public static double BotRadarTurnRate = 34.1;
    public static double BotGunHeat = 7.6;

    private double _botEnergy = BotEnergy;
    private double _botGunHeat = BotGunHeat;

    private WebSocketServer _server;

    private BotHandshake _botHandshake;
    private BotIntent _botIntent;

    private int _turnNumber = 1;

    private readonly EventWaitHandle _openedEvent = new AutoResetEvent(false);
    private readonly EventWaitHandle _botHandshakeEvent = new AutoResetEvent(false);
    private readonly EventWaitHandle _gameStartedEvent = new AutoResetEvent(false);
    private readonly EventWaitHandle _tickEvent = new AutoResetEvent(false);
    private readonly EventWaitHandle _botIntentEvent = new AutoResetEvent(false);

    public static Uri ServerUrl => new($"ws://127.0.0.1:{Port}");

    public void Start()
    {
        _server = new WebSocketServer(ServerUrl.AbsoluteUri, false);
        _server.Start(conn =>
        {
            conn.OnOpen = () => OnOpen(conn);
            conn.OnMessage = message => OnMessage(conn, message);
            conn.OnError = OnError;
        });
    }

    public void Stop()
    {
        _server.Dispose();
        // TODO: Sleep for 100 millis?
    }

    public void SetBotEnergy(double botEnergy)
    {
        _botEnergy = botEnergy;
    }

    public void SetBotGunHeat(double botGunHeat)
    {
        _botGunHeat = botGunHeat;
    }

    public bool AwaitConnection(int milliSeconds)
    {
        try
        {
            return _openedEvent.WaitOne(milliSeconds);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("AwaitConnection: Exception occurred: " + ex);
        }

        return false;
    }

    public bool AwaitBotHandshake(int milliSeconds)
    {
        try
        {
            return _botHandshakeEvent.WaitOne(milliSeconds);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("AwaitBotHandshake: Exception occurred: " + ex);
        }

        return false;
    }

    public bool AwaitGameStarted(int milliSeconds)
    {
        try
        {
            return _gameStartedEvent.WaitOne(milliSeconds);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("AwaitGameStarted: Exception occurred: " + ex);
        }

        return false;
    }

    public bool AwaitTickEvent(int milliSeconds)
    {
        try
        {
            return _tickEvent.WaitOne(milliSeconds);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("AwaitTickEvent: Exception occurred: " + ex);
        }

        return false;
    }

    public bool AwaitBotIntent(int milliSeconds)
    {
        try
        {
            return _botIntentEvent.WaitOne(milliSeconds);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("AwaitBotIntentEvent: Exception occurred: " + ex);
        }

        return false;
    }

    public BotHandshake GetBotHandshake() => _botHandshake;


    private void OnOpen(IWebSocketConnection conn)
    {
        _openedEvent.Set();
        SendServerHandshake(conn);
    }

    private void OnMessage(IWebSocketConnection conn, string messageJson)
    {
        Console.WriteLine("OnMessage: " + messageJson);

        var message = JsonConvert.DeserializeObject<Message>(messageJson);
        if (message == null) return;

        var msgType = (MessageType)Enum.Parse(typeof(MessageType), message.Type);
        switch (msgType)
        {
            case MessageType.BotHandshake:
                _botHandshake = JsonConvert.DeserializeObject<BotHandshake>(messageJson);
                _botHandshakeEvent.Set();

                SendGameStartedForBot(conn);
                _gameStartedEvent.Set();
                break;

            case MessageType.BotReady:
                SendRoundStarted(conn);

                SendTickEventForBot(conn, _turnNumber++);
                _tickEvent.Set();
                break;

            case MessageType.BotIntent:
                _botIntentEvent.Set();

                _botIntent = JsonConvert.DeserializeObject<BotIntent>(messageJson);
                Thread.Sleep(5);

                SendTickEventForBot(conn, _turnNumber++);
                break;
        }
    }

    private static void OnError(Exception ex)
    {
        throw new InvalidOperationException("MockedServer error", ex);
    }

    private void SendServerHandshake(IWebSocketConnection conn)
    {
        var serverHandshake = new ServerHandshake
        {
            Type = EnumUtil.GetEnumMemberAttrValue(MessageType.ServerHandshake),
            SessionId = SessionId,
            Name = Name,
            Version = Version,
            Variant = Variant,
            GameTypes = GameTypes
        };
        Send(conn, serverHandshake);
    }

    private void SendGameStartedForBot(IWebSocketConnection conn)
    {
        var gameStarted = new GameStartedEventForBot
        {
            Type = EnumUtil.GetEnumMemberAttrValue(MessageType.GameStartedEventForBot),
            MyId = MyId
        };
        var gameSetup = new Schema.GameSetup
        {
            GameType = GameType,
            ArenaWidth = ArenaWidth,
            ArenaHeight = ArenaHeight,
            NumberOfRounds = NumberOfRounds,
            GunCoolingRate = GunCoolingRate,
            MaxInactivityTurns = MaxInactivityTurns,
            TurnTimeout = TurnTimeout,
            ReadyTimeout = ReadyTimeout
        };
        gameStarted.GameSetup = gameSetup;
        Send(conn, gameStarted);
    }

    private void SendRoundStarted(IWebSocketConnection conn)
    {
        var roundStarted = new RoundStartedEvent
        {
            Type = EnumUtil.GetEnumMemberAttrValue(MessageType.RoundStartedEvent),
            RoundNumber = 1
        };
        Send(conn, roundStarted);
    }

    private void SendTickEventForBot(IWebSocketConnection conn, int turnNumber)
    {
        var tickEvent = new TickEventForBot()
        {
            Type = EnumUtil.GetEnumMemberAttrValue(MessageType.TickEventForBot),
            RoundNumber = 1,
            TurnNumber = turnNumber,
            EnemyCount = BotEnemyCcount
        };

        var turnRate = BotTurnRate;
        var gunTurnRate = BotGunTurnRate;
        var radarTurnRate = BotRadarTurnRate;

        if (_botIntent != null)
        {
            turnRate = _botIntent.TurnRate ?? BotTurnRate;
            gunTurnRate = _botIntent.GunTurnRate ?? BotGunTurnRate;
            radarTurnRate = _botIntent.RadarTurnRate ?? BotRadarTurnRate;
        }

        var state = new Schema.BotState
        {
            Energy = _botEnergy,
            X = BotX,
            Y = BotY,
            Direction = BotDirection,
            GunDirection = BotGunDirection,
            RadarDirection = BotRadarDirection,
            RadarSweep = BotRadarSweep,
            Speed = BotSpeed,
            TurnRate = turnRate,
            GunTurnRate = gunTurnRate,
            RadarTurnRate = radarTurnRate,
            GunHeat = _botGunHeat
        };
        tickEvent.BotState = state;

        var bulletState1 = CreateBulletState(1);
        var bulletState2 = CreateBulletState(2);
        tickEvent.BulletStates = new HashSet<Schema.BulletState>
        {
            bulletState1, bulletState2
        };

        var scannedEvent = new ScannedBotEvent()
        {
            Type = EnumUtil.GetEnumMemberAttrValue(MessageType.ScannedBotEvent),
            Direction = 45,
            X = 134.56,
            Y = 256.7,
            Energy = 56.9,
            Speed = 9.6,
            TurnNumber = 1,
            ScannedBotId = 2,
            ScannedByBotId = 1
        };
        tickEvent.Events = new HashSet<Event> { scannedEvent };

        Send(conn, tickEvent);
    }

    private void Send(IWebSocketConnection conn, Object obj)
    {
        conn.Send(JsonConvert.SerializeObject(obj));
    }

    private static Schema.BulletState CreateBulletState(int id)
    {
        var bulletState = new Schema.BulletState()
        {
            BulletId = id,
            X = 0,
            Y = 0,
            OwnerId = 0,
            Direction = 0,
            Power = 0
        };
        return bulletState;
    }
}